rclcpp/include/rclcpp/experimental/subscription_intra_process_buffer.hpp

```diff
@@ -163,6 +164,67 @@ class SubscriptionIntraProcessBuffer : public SubscriptionROSMsgIntraProcessBuff
     return buffer_->use_take_shared_method();
   }
 
+  bool
+  is_serialized() const override
+  {
+    return serialization_traits::is_serialized_message_class<SubscribedType>::value;
+  }
+
+  bool serve_serialized_message(
+    const rclcpp::SerializedMessage * serialized_message,
+    IntraProcessManager * intraprocess_manager,
+    uint64_t intra_process_publisher_id,
+    void * untyped_allocator,
+    const std::vector<uint64_t> & take_ownership_subscriptions,
+    const std::vector<uint64_t> & take_shared_subscriptions) override
+  {
+    if (intraprocess_manager != nullptr && serialized_message != nullptr) {
+      typedef typename allocator::AllocRebind<SubscribedType, Alloc>::allocator_type AllocType;
+      auto allocator = static_cast<AllocType *>(untyped_allocator);
+
+      if (allocator == nullptr) {
+        throw std::runtime_error(
+                "failed to cast allocator "
+                "which can happen when the publisher and "
+                "subscription use different allocator types, which is not supported");
+      }
+
+      auto ptr = SubscribedTypeAllocatorTraits::allocate(subscribed_type_allocator_, 1);
+      SubscribedTypeAllocatorTraits::construct(subscribed_type_allocator_, ptr);
+      if constexpr (std::is_same<SubscribedType,
+        rclcpp::SerializedMessage>::value ||
+        std::is_same<SubscribedType, rclcpp::SerializedMessage>::value)
+      {
+        (void)serialized_message;
+        throw std::runtime_error("Serialized message cannot be serialized.");
+      } else if constexpr (std::is_same<SubscribedType, ROSMessageType>::value) {
+        Serialization<SubscribedType> serializer;
+        serializer.deserialize_message(serialized_message, ptr);
+      } else {
+        ROSMessageType ros_msg;
+        Serialization<ROSMessageType> serializer;
+        serializer.deserialize_message(serialized_message, &ros_msg);
+        rclcpp::TypeAdapter<SubscribedType, ROSMessageType>::convert_to_custom(ros_msg, *ptr);
+      }
+      SubscribedTypeUniquePtr message(ptr, subscribed_type_deleter_);
+
+      detail::do_intra_process_publish_same_type<SubscribedType, ROSMessageType, Alloc, Deleter>(
```

Comment:
I'm sorry, but I have been out of the  loop with the recent changes to the intra-process manager, so maybe may question is not related to this PR.
This class is a subscription buffer: why does  it call a "publish" function?

Reply:
from comment:

 * Helper function to expose method of IntraProcessManager
 * for publishing messages with same data type.
 *
 * This is needed as the publisher of serialized message is not aware of the subscribed
 * data type. While the subscription has all needed information (MessageT, Allocator) to
 * cast and deserialize the message. The type information is forwarded by the helper function.

Reply:
I think that this helper function should be modified.
It currently does two separate things:
 - create a message of a certain type
 - invoke intra-process publish

IMO the helper should just take care of creating the message.
Then it should return it as output such that intra-process publish is invoked elsewhere.

Reply:
I've splitted it into two methods (one for deserialization and one for forwarding). Yet, the message type is only known within this class so it has to stay here (alternative would be generating a lambda function).